### PR TITLE
Multiples Update

### PR DESCRIPTION
**Pull request by vsc55**
Hi @madgen78,

Thanks for the plugin it is very useful.

Here I'm sending you a PR with a few updates to see what you think.

Data collection has now been segmented into different files, one for each table or destination.
Remove folder graphviz and installa via composer.
Applying patches in graphviz using cweagans.
Create i18n translation to es_ES.
Remove install.php, insert in Class Module.
Remove functions.inc.php, insert in Class Module.
Extract function dpp in new class.
Change mode save Settings, now save in kvstore (settings are lost when updating!!!)
Add button to reset default settgins.
Remove setting destination, now get destination by ajax.
Update rnav.
Implement backup and restore.

### CODE REVIEW

**Comment by madgen78** [comment]
I like your changes. It updates and makes it more modern.
Installed main from your repo causes some issues.
time conditions fail  Undefined variable $tgLabel 

image
manually setting the following creates a new error with ivr: Undefined array key "announcement"
$tgLabel='';
$tgLink='';
$tgTooltip='';

image
manually setting 
$ivr['name']='';
$ivr['invalid_destination'] ='';
$ivr['timeout_destination']='';

does draw the graph.
Extension nodes from queues are white instead of green.

I'm working on combining all extension destinations into one node instead of just queues and ring groups. It's not perfect yet. (External numbers should not say "Extension: ")
I've changed all nodes to be shape=rect.
Here is how it would look.

ANY (6)


By using rect, the graph is more compact and does not create extra spacing for the different shapes.
Because it is more compact, the lines and labels are closer together. Hovering over the edge labels says where it came from and highlights the path.
The shapes were in place to identify the node. Now, the color will have to do the same thing.
I would like to give the user the ability to change shapes and colors, but this sounds like a lot of work.

Back to your version...
I like your idea of keeping the settings in kvstore as updating resets this.
I like the Settings tab instead of a dropdown and the 3 combined buttons instead of 3 separate.
The navigation section shouldn't have an x to close it. Maybe a third tab instead?
I did have the same idea to load the graph with ajax instead of reloading the page. Haven't gotten around to it yet.
The destination checkbox was not easy to find and thought that you removed it.
-are you loading these by default? or does it query when checking the check box?
--on larger installations (300 + inbound routes) ...loading destinations by default, lags the page each time. Disabling the feature loads much quicker.

Thanks for your help on this.

**Comment by vsc55** [comment]
>  -are you loading these by default? or does it query when checking the check box?
Destinations are loaded dynamically in the background.
If you don't have the destination column selected, that data won't be loaded. However, when the destination column is marked as visible, it is loaded in the background via Ajax.
https://github.com/vsc55/freepbx_dpviz/blob/ee44d0f7e8b8cc2fe0a5042569bb2a1e3bef47e4/assets/js/Dpviz.js#L215

**Comment by madgen78** [comment]
FYI...I've just added a single update.


```
foreach($results as $users) {
		$Qresult=array();
    $id = $users['extension'];
    $u[$id]= $users;
    $dproute['extensions'][$id]= $users;
		$emailResult=[];    <----new. This starts with an empty array instead of adding. Only the first email address would be correct.
		$email='grep -E \'^'.$id.'[[:space:]]*[=>]+\' /etc/asterisk/voicemail.conf | cut -d \',\' -f3';
		exec($email, $emailResult);
		if (!empty($emailResult[0])){
			$dproute['extensions'][$id]['email'] =$emailResult[0];
		}else{
			$dproute['extensions'][$id]['email'] ='unassigned';
		}
  }
```

**Comment by vsc55** [comment]
> FYI...I've just added a single update.
> 
> ```
> foreach($results as $users) {
>     $emailResult=[];    <----new. This starts with an empty array instead of adding. Only the first email address would be correct.
> 
>     $id = $users['extension'];
>     $u[$id]= $users;
>     $dproute['extensions'][$id]= $users;
> 		$email='grep -E \'^'.$id.'[[:space:]]*[=>]+\' /etc/asterisk/voicemail.conf | cut -d \',\' -f3';
> 		exec($email, $emailResult);
> 		if (!empty($emailResult[0])){
> 			$dproute['extensions'][$id]['email'] =$emailResult[0];
> 		}else{
> 			$dproute['extensions'][$id]['email'] ='unassigned';
> 		}
>   }
> ```

Why do you define this if it is not used?
`$u[$id]= $users;`

**Comment by madgen78** [comment]
Thanks. Leftover from prior versions. I've removed it.

**Comment by vsc55** [comment]
> Extension nodes from queues are white instead of green.

I think the problem is here, I had it marked as a possible error.
`
if ($route['parent_edge_label'] == _(' Static'))
{
    $node->attribute('fillcolor', self::pastels[20]);
}
else
{
    $node->attribute('fillcolor', self::pastels[8]);
}
`
The problem is that we use "$route['parent edge label']" to see if it is static or dynamic, and when translating this text it is giving some error.
I'm going to look into it and see what can be done.

**Comment by vsc55** [comment]
> > Extension nodes from queues are white instead of green.
> 
> I think the problem is here, I had it marked as a possible error.
> 
> ```
> if ($route['parent_edge_label'] == _(' Static'))
> {
>     $node->attribute('fillcolor', self::pastels[20]);
> }
> else
> {
>     $node->attribute('fillcolor', self::pastels[8]);
> }
> ```
> 
> The problem is that we use "$route['parent edge label']" to see if it is static or dynamic, and when translating this text it is giving some error. I'm going to look into it and see what can be done.

In the end it wasn't that, it's because of a change you made in version .17.
https://github.com/madgen78/dpviz/commit/96afee919cfb3054d8b1b51f769a5fb263f99d84#diff-2b1ac0ec5796aa0b945b800ca75e6179372f7718ca01b8d6a72921eaaa271628L764

**Comment by vsc55** [comment]
What do you think of this toolbar:
imagen
I have also added an input so you can define the name you want for the file.

**Comment by madgen78** [comment]
yea.. I like that idea. maybe too wide. maybe have the placeholder the phone number as default if not changed.

**Comment by madgen78** [comment]
finished version- thanks for the suggestion! tested on fpbx14,16,17
The name is a placeholder, can be modified. script adds  .png to the filename after choosing high or standard.
image

**Comment by vsc55** [comment]
What do you think of this style, to separate the export buttons from the ones that are for working with the diagram.
imagen

**Comment by vsc55** [comment]
@madgen78 have you tried svg-pan-zoom?
https://github.com/bumbu/svg-pan-zoom

**Comment by vsc55** [comment]
Coming soon, menu to select the "Inbound Routes" in the toolbar 😉
imagen

**Comment by vsc55** [comment]
New dialplane selector:

https://github.com/user-attachments/assets/d5435d92-8ecd-4d03-8bb4-b672da7b0e5f

**Comment by miken32** [comment]
IMO from a code perspective these changes look good – it seems much less like a module from 2.9 days!

Changes I would suggest are PSR-12 conformance (this would make things much easier for other contributors) and consistent and detailed typing of method parameters and class properties. Like, looking at the code for the `FreePBX\Modules\Dpviz\dpp\tables\baseTables` class, I don't know what `$this->dpp` is, beyond that it's an object. Or, none of the methods in `FreePBX\Modules\Dpviz\dpp\destination\baseDestination` have typed parameters at all.

Also, `PDO::fetchAll()` can index by a column using `PDO::FETCH_KEY_PAIR`, it would save you a few dozen loops.
